app/src/main/java/de/danoeh/antennapod/fragment/QueueFragment.java

```diff
@@ -325,37 +339,48 @@ public void onConfirmButtonPressed(
-                    QueueSorter.sort(getActivity(), QueueSorter.Rule.EPISODE_TITLE_DESC, true);
+                    setSortOrder(SortOrder.EPISODE_TITLE_Z_A);
                     return true;
                 case R.id.queue_sort_date_asc:
-                    QueueSorter.sort(getActivity(), QueueSorter.Rule.DATE_ASC, true);
+                    setSortOrder(SortOrder.DATE_OLD_NEW);
                     return true;
                 case R.id.queue_sort_date_desc:
-                    QueueSorter.sort(getActivity(), QueueSorter.Rule.DATE_DESC, true);
+                    setSortOrder(SortOrder.DATE_NEW_OLD);
                     return true;
                 case R.id.queue_sort_duration_asc:
-                    QueueSorter.sort(getActivity(), QueueSorter.Rule.DURATION_ASC, true);
+                    setSortOrder(SortOrder.DURATION_SHORT_LONG);
                     return true;
                 case R.id.queue_sort_duration_desc:
-                    QueueSorter.sort(getActivity(), QueueSorter.Rule.DURATION_DESC, true);
+                    setSortOrder(SortOrder.DURATION_LONG_SHORT);
                     return true;
                 case R.id.queue_sort_feed_title_asc:
-                    QueueSorter.sort(getActivity(), QueueSorter.Rule.FEED_TITLE_ASC, true);
+                    setSortOrder(SortOrder.FEED_TITLE_A_Z);
                     return true;
                 case R.id.queue_sort_feed_title_desc:
-                    QueueSorter.sort(getActivity(), QueueSorter.Rule.FEED_TITLE_DESC, true);
+                    setSortOrder(SortOrder.FEED_TITLE_Z_A);
                     return true;
                 case R.id.queue_sort_random:
-                    QueueSorter.sort(getActivity(), QueueSorter.Rule.RANDOM, true);
+                    setSortOrder(SortOrder.RANDOM);
                     return true;
                 case R.id.queue_sort_smart_shuffle_asc:
-                    QueueSorter.sort(getActivity(), QueueSorter.Rule.SMART_SHUFFLE_ASC, true);
+                    setSortOrder(SortOrder.SMART_SHUFFLE_OLD_NEW);
                     return true;
                 case R.id.queue_sort_smart_shuffle_desc:
-                    QueueSorter.sort(getActivity(), QueueSorter.Rule.SMART_SHUFFLE_DESC, true);
+                    setSortOrder(SortOrder.SMART_SHUFFLE_NEW_OLD);
+                    return true;
+                case R.id.queue_keep_sorted:
+                    boolean keepSortedOld = UserPreferences.isQueueKeepSorted();
+                    boolean keepSortedNew = !keepSortedOld;
+                    UserPreferences.setQueueKeepSorted(keepSortedNew);
+                    if (keepSortedNew) {
+                        SortOrder sortOrder = UserPreferences.getQueueKeepSortedOrder();
+                        QueueSorter.sort(sortOrder, true);
+                    }
+                    // Update sort menu items and list lock elements
+                    getActivity().recreate();
```

Comment:
I think that's a bit too much (re-binding to services etc). This should be enough:

```
getActivity().invalidateOptionsMenu();
recyclerAdapter.notifyDataSetChanged();
```

Reply:
My suggestion does not work. I suggest this instead:

```
if (keepSortedNew) {
    SortOrder sortOrder = UserPreferences.getQueueKeepSortedOrder();
    QueueSorter.sort(sortOrder, true);
    recyclerAdapter.setLocked(true);
} else {
    recyclerAdapter.setLocked(UserPreferences.isQueueLocked());
}
getActivity().invalidateOptionsMenu();
```

Reply:
I faced the same problem when I tried `recyclerAdapter.notifyDataSetChanged()` to update the list. But your suggestion looks much better.

Reply:
Now it works :thumbsup: